src/infra/bus/distributed.js

```diff
@@ -0,0 +1,184 @@
+      payload: JSON.stringify(payload),
+    };
+    await this.#redis.xAdd(streamKey, '*', data, {
+      TRIM: { strategy: 'MAXLEN', strategyModifier: '~', threshold: 10000 },
+    });
+    return true;
+  }
+
+  /** @type IPubSub['subscribe'] */
+  subscribe(event, handler) {
+    this.#eventHandlers.set(event, handler);
+    return true;
+  }
+
+  /** @type IPubSub['unsubscribe'] */
+  unsubscribe(event) {
+    this.#eventHandlers.delete(event);
+    return true;
+  }
+
+  /** @type ICommand['call'] */
+  async call({ service: serviceName, method }, payload) {
+    const callId = randomUUID();
+    const streamKey = `${serviceName}:${method}:request`;
+    /** @type CallData */
+    const data = {
+      serverId: this.#serverId,
+      callId,
+      payload: JSON.stringify(payload),
+    };
+    await this.#redis.xAdd(streamKey, '*', data, {
+      TRIM: { strategy: 'MAXLEN', strategyModifier: '~', threshold: 10000 },
+    });
+    return new Promise((resolve, reject) => {
+      this.#calls.set(callId, { resolve, reject });
+    });
+  }
+
+  /** @type {(message: string, channel: string) => void} */
+  #handleResponse(message, channel) {
+    const callId = channel.split(':').at(-1);
+    const promise = this.#calls.get(callId);
+    if (!promise) return; // need to handle properly
+    /** @type CallResponse */
+    const response = JSON.parse(message);
+    promise.resolve(response);
+  }
+
+  /** @type {(serviceName: string) => Promise<void>} */
+  async #listenRemoteCalls(serviceName) {
+    const service = this.#services.get(serviceName);
+    const streams = [];
+    for (const methodName of Object.keys(service)) {
+      const key = `${serviceName}:${methodName}:request`;
+      streams.push({ key, id: '>' });
+    }
+    const events = await this.#redis.xReadGroup(
+      serviceName,
+      this.#serverId,
+      streams,
+      { COUNT: 1, BLOCK: 10000, NOACK: true },
+    );
+    if (!events?.length) return;
+    const { name: streamName, messages } = events[0];
+    const { message } = messages[0];
+    const payload = /** @type CallData */ (message);
+    this.#handleRemoteCall(streamName, payload);
+  }
+
+  /** @type {(streamName: string, data: CallData) => Promise<void>} */
+  async #handleRemoteCall(
+    streamName,
+    { serverId, callId, payload: strPayload },
+  ) {
+    const [service, method] = streamName.split(':');
+    const payload = JSON.parse(strPayload);
+
+    const result = await this.#localCall({ service, method }, payload);
+
+    const message = JSON.stringify(result);
+    this.#redis.publish(`response:${serverId}:${callId}`, message);
+  }
+
+  /** @type ICommand['call'] */
+  async #localCall({ service: serviceName, method }, payload) {
+    const service = this.#services.get(serviceName);
+    const handler = service[method];
+    if (!handler)
+      return [{ expected: false, message: 'Method not found' }, null];
+    const result = await handler(payload);
+    return result;
+  }
```

Comment:
Why you need local call here?

Reply:
To handle receiving remote commands.
Basically, it's just a helper function to extract the "passing remote call data to local service" logic.
I think it will be easier to maintain buses since `LocalBus` has the same function.